pkg/lightning/backend/local/duplicate.go

```diff
@@ -0,0 +1,636 @@
+				} else {
+					unfinishedRegions = append(unfinishedRegions, r)
+				}
+			} else {
+				waitingClients = append(waitingClients, cli)
+				watingRegions = append(watingRegions, region)
+			}
+		}
+
+		if len(indexHandles) > 0 {
+			handles := manager.getValues(ctx, indexHandles)
+			if len(handles) > 0 {
+				indexHandles = handles
+			} else {
+				indexHandles = indexHandles[:0]
+			}
+		}
+
+		for idx, cli := range waitingClients {
+			region := watingRegions[idx]
+			for {
+				resp, reqErr := cli.Recv()
+				hasErr := false
+				if reqErr != nil {
+					if errors.Cause(reqErr) == io.EOF {
+						break
+					} else {
+						hasErr = true
+					}
+				}
+
+				if hasErr || resp.GetKeyError() != nil {
+					r, err := manager.splitCli.GetRegionByID(ctx, region.Region.GetId())
+					if err != nil {
+						unfinishedRegions = append(unfinishedRegions, region)
+					} else {
+						unfinishedRegions = append(unfinishedRegions, r)
+					}
+				}
+				if hasErr {
+					log.L().Warn("meet error when recving duplicate detect response from TiKV, retry again",
+						logutil.Region(region.Region), logutil.Leader(region.Leader), zap.Error(reqErr))
+					break
+				}
+				if resp.GetKeyError() != nil {
+					log.L().Warn("meet key error in duplicate detect response from TiKV, retry again ",
+						logutil.Region(region.Region), logutil.Leader(region.Leader),
+						zap.String("KeyError", resp.GetKeyError().GetMessage()))
+					break
+				}
+
+				if resp.GetRegionError() != nil {
+					log.L().Warn("meet key error in duplicate detect response from TiKV, retry again ",
+						logutil.Region(region.Region), logutil.Leader(region.Leader),
+						zap.String("RegionError", resp.GetRegionError().GetMessage()))
+
+					r, err := paginateScanRegion(ctx, manager.splitCli, watingRegions[idx].Region.GetStartKey(), watingRegions[idx].Region.GetEndKey(), 1024)
+					if err != nil {
+						unfinishedRegions = append(unfinishedRegions, watingRegions[idx])
+					} else {
+						unfinishedRegions = append(unfinishedRegions, r...)
+					}
+					break
+				}
+
+				handles, err := manager.storeDuplicateData(ctx, resp, decoder, req)
+				if err != nil {
+					return err
+				}
+				if handles != nil && len(handles) > 0 {
+					indexHandles = append(indexHandles, handles...)
+				}
+			}
+		}
+
+		// it means that all of region send to TiKV fail, so we must sleep some time to avoid retry too frequency
+		if len(unfinishedRegions) == len(regions) {
+			tryTimes += 1
+			time.Sleep(defaultRetryBackoffTime)
+		}
+		regions = unfinishedRegions
+	}
+	return nil
+}
+
+func (manager *DuplicateManager) storeDuplicateData(
+	ctx context.Context,
+	resp *sst.DuplicateDetectResponse,
+	decoder *backendkv.TableKVDecoder,
+	req *DuplicateRequest,
+) ([][]byte, error) {
+	opts := &pebble.WriteOptions{Sync: false}
+	var err error
+	totalKeyLen := 0
+	for _, kv := range resp.Pairs {
+		totalKeyLen += manager.keyAdapter.EncodedLen(kv.Key)
+	}
+	buf := make([]byte, totalKeyLen)
```

Comment:
It seems `buf` is only used for a single key. It's unnecessary to create a slice with total key length.

Reply:
done